### PR TITLE
feat(publisher): pass commit_id to Summary Report for Redis dedup idempotency (#3253)

### PR DESCRIPTION
**Pull request by RC918**
# feat(publisher): pass commit_id to Summary Report for Redis dedup idempotency (#3253)

## Summary
Fixes issue #3253 where Summary Report and file-level fallback paths in `publisher_node` were not passing `commit_id` to `post_pr_review()`, causing the Redis dedup mechanism to be bypassed. This could result in duplicate Summary Reports when the Reviewer Agent is re-run on the same PR.

**Changes:**
- Pass `commit_id=stored_head_sha` to `post_pr_review()` in Summary Report path
- Pass `commit_id=stored_head_sha` to `post_pr_review()` in file-level fallback path  
- Add defensive validation to normalize `diff_head_sha` to `None` if not a valid non-empty string (applied consistently to both Summary Report path and inline comments path)
- Add tests to verify `commit_id` is passed for Redis dedup and edge cases are handled

## Updates since last revision
- **Unified defensive validation**: Per gemini-code-assist feedback, added the same defensive validation to the inline comments path (line 5731) that was already in the Summary Report path. Both paths now use:
  ```python
  _raw_head_sha = state.get("diff_head_sha")
  stored_head_sha = _raw_head_sha if isinstance(_raw_head_sha, str) and _raw_head_sha else None
  ```

- **Added file-level fallback test**: Added `test_file_level_fallback_passes_commit_id_for_dedup` to verify `commit_id` is correctly passed in the file-level fallback path (when comments exist but none are inline-eligible).

- **Created follow-up issues** for suggestions that were out of scope:
  - #3258: Integration test for Redis dedup verification
  - #3259: Define `diff_head_sha` contract and ensure availability
  - #3260: Add observability for Redis dedup failures

## Review & Testing Checklist for Human
- [ ] **Verify defensive validation consistency**: Both paths (Summary Report at line ~5633 and inline comments at line ~5733) should have identical validation logic
- [ ] **Test idempotency**: Re-run the Reviewer Agent on the same PR twice and verify only one Summary Report is posted (requires Redis and `GITHUB_REVIEW_POSTING_DRY_RUN=false`)
- [ ] **Test graceful degradation**: Verify behavior when `diff_head_sha` is missing from state (should proceed without dedup, not crash)

### Notes
- The regular inline comments path already passes `commit_id` correctly (line ~5996), so this PR aligns the Summary Report and file-level fallback paths with that pattern
- When `stored_head_sha` is `None`, `post_pr_review()` handles it gracefully (dedup is skipped but review still posts)
- Strict SHA format validation (40-hex regex) was intentionally not added to avoid churn with existing test fixtures; tracked in follow-up issue #3259

---
Requested by: Ryan Chen (@RC918)
Link to Devin run: https://app.devin.ai/sessions/4c624f57c50e499f98859c62908c800d

### CODE REVIEW

**Review by RC918 — COMMENTED** [review]
## MorningAI Code Review

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 1]
```diff
@@ -5627,12 +5627,17 @@ def publisher_node(state: AgentState) -> AgentState:
 """
 
             repo = get_repo()
+            # Issue #3253: Get commit_id for Redis dedup idempotency
+            # Normalize to None if not a valid non-empty string (defensive)
+            raw_head_sha = state.get("diff_head_sha")
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: The value from `state.get("diff_head_sha")` is assumed to be a commit SHA, but no validation is performed to confirm it's a valid Git SHA format. If malformed data enters this path, it could lead to inconsistent Redis keys or failed lookups. SUGGESTED FIX: Add basic format validation (e.g., regex for 40-character hex string) or sanitize before use as `commit_id`.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 2]
```diff
@@ -5627,12 +5627,17 @@ def publisher_node(state: AgentState) -> AgentState:
...
+            stored_head_sha = raw_head_sha if isinstance(raw_head_sha, str) and raw_head_sha else None
             if repo:
                 result = post_pr_review(
                     repo=repo,
                     pr_number=pr_number,
                     comments=[],
-                    summary=summary_body
+                    summary=summary_body,
+                    commit_id=stored_head_sha  # Enable Redis dedup
```
Comment (RC918) [inline handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: IMPACT: Passing `None` as `commit_id` may bypass deduplication logic on the receiving end if not handled explicitly, potentially leading to duplicates. SUGGESTED FIX: Consider whether `post_pr_review()` should reject or default to another identifier (e.g., PR number + timestamp) when `commit_id` is `None`, and document expected behavior.

**Review by RC918 — COMMENTED** [review]
## MorningAI Code Review

File: handoff/20250928/40_App/orchestrator/tests/test_publisher_node.py [hunk 3]
```diff
@@ -744,6 +744,44 @@ def test_no_comments_at_all_posts_summary_report(self):
                         # Summary Report uses "MorningAI Review Summary" as the header
                         assert "MorningAI Review Summary" in call_args.kwargs.get("summary", "")
 
+    def test_summary_report_passes_commit_id_for_dedup(self):
```
[the remark this answers is not in the document]
Comment (RC918) [reply handoff/20250928/40_App/orchestrator/tests/test_publisher_node.py]: Added test case in commit 45234dd8 covering missing/invalid `diff_head_sha` values (None, empty string, int, dict, list). The test verifies that `commit_id` is normalized to `None` for all invalid values.

File: handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py [hunk 4]
```diff
@@ -5627,12 +5627,15 @@ def publisher_node(state: AgentState) -> AgentState:
 """
 
             repo = get_repo()
+            # Issue #3253: Get commit_id for Redis dedup idempotency
+            stored_head_sha = state.get("diff_head_sha")
```
[the remark this answers is not in the document]
Comment (RC918) [reply handoff/20250928/40_App/orchestrator/langgraph_orchestrator.py]: Addressed in commit 4abbc292. Added defensive validation to the inline comments path (line 5731) to match the Summary Report path. Now both paths use the same pattern:
```python
_raw_head_sha = state.get("diff_head_sha")
stored_head_sha = _raw_head_sha if isinstance(_raw_head_sha, str) and _raw_head_sha else None
```

This ensures consistent type safety across all `post_pr_review()` calls.
